src/homebridge-roku.js

```diff
@@ -47,35 +66,93 @@ class RokuAccessory {
     accessoryInfo
       .setCharacteristic(Characteristic.Manufacturer, this.info.vendorName)
       .setCharacteristic(Characteristic.Model, this.info.modelName)
-      .setCharacteristic(Characteristic.Name, this.info.userDeviceName)
       .setCharacteristic(Characteristic.SerialNumber, this.info.serialNumber);
 
+    this.services.push(accessoryInfo);
     return accessoryInfo;
   }
 
-  setupSwitch() {
-    const switch_ = new Service.Switch(`${this.name} Power`, 'Power');
+  setupTelevision() {
+    const television = new Service.Television(this.name);
 
-    switch_
-      .getCharacteristic(Characteristic.On)
-      .on('get', callback => callback(null, this.poweredOn))
-      .on('set', (value, callback) => {
-        this.poweredOn = value;
+    television
+      .getCharacteristic(Characteristic.Active)
+      .on('get', (callback) => {
+        this.roku
+          .info()
+          .then((info) => {
+            const value = info.powerMode === 'PowerOn' ? Characteristic.Active.ACTIVE : Characteristic.Active.INACTIVE;
+            this.poweredOn = value;
+            callback(null, value);
+          }, callback);
+      })
+      .on('set', (newValue, callback) => {
+        console.log('Wants to power on, currently is', this.poweredOn)
+        if (newValue == this.poweredOn) {
+          callback(null);
+          return;
+        }
+        this.poweredOn = newValue;
         this.roku
           .keypress('Power')
-          .then(() => callback(null))
-          .catch(callback);
+          .then(() => callback(null), callback);
+      })
+
+    television
+      .getCharacteristic(Characteristic.ActiveIdentifier)
+      .on('get', (callback) => {
+        this.roku
+          .active()
+          .then((app) => {
+            const index = app !== null ? this.inputs.findIndex((input) => input.id == app.id) : -1;
+            const hapID = index + 1;
+            callback(null, hapID);
+          }, callback);
+      })
+      .on('set', (index, callback) => {
+        const rokuID = this.inputs[index - 1].id
+        this.roku
+          .launch(rokuID)
+          .then(() => callback(null), callback);
+      })
+
+    television
+      .getCharacteristic(Characteristic.ConfiguredName)
+      .setValue(this.info.userDeviceName)
+      .setProps({
+        perms: [Characteristic.Perms.READ]
+      });
+
+    television
+      .setCharacteristic(Characteristic.SleepDiscoveryMode, Characteristic.SleepDiscoveryMode.ALWAYS_DISCOVERABLE);
```

Comment:
What does this do? I know my TV is not discoverable from sleep, it turns off its network interface.

Reply:
No clue. I can try binding this to one of the config values. How do ever turn yours back on?

Reply:
Newer TCLs stay available when they sleep. My 2018 series 6 allows for me to turn it on from the app.

Reply:
Right - mine's the same kind of TCL. I don't know what values I should be looking for in `device-info`, nor what overall to do about _the other kind_ that apparently shuts off its NIC.

Reply:
I have to turn it on with my remote :P There's [this issue](https://github.com/bschlenk/homebridge-roku/issues/9) where a few configs are posted, but nothing jumps out at me as meaning it can be turned on from sleep. It probably wouldn't hurt anything to leave this set to `ALWAYS_DISCOVERABLE` - do you know if this changes anything in the iOS home app or remote interface?

Reply:
`NOT_DISCOVERABLE` might mean it doesn't enter a failure state if it can't find it immediately. I can try it…